### PR TITLE
Correct intent for "GH_WRITE" to "inout" in kernels (for #1003)

### PR DESCRIPTION
**Pull request by TeranIvy**
According to Fortran rules the field and operator data arguments in kernels should have `intent(inout)` as they are always defined outside of kernels. However, we currently generate `intent(out)` in LFRic kernels for arguments that have `GH_WRITE` access. This PR corrects the generation and access rules for `GH_WRITE` access to `intent(inout)`. 

P.S. I checked the coverage following the link ["Continue to review full report at Codecov"](https://codecov.io/gh/stfc/PSyclone/pull/1117?src=pr&el=continue) from below and everything there seems accounted for.

### CODE REVIEW

**Comment by rupertford** [comment]
Hi @TeranIvy, I thought that if basic types were passed in to a subroutine and were only written then they should be intent(out). What am I missing?

**Comment by TeranIvy** [comment]
> 
> 
> Hi @TeranIvy, I thought that if basic types were passed in to a subroutine and were only written then they should be intent(out). What am I missing?

@rupertford, the whole of a field array is passed into a kernel, but the kernel only modifies one of the columns. Some compilers with some compile options will overwrite the full array of `intent(out)`, so the intent needs to be `inout`. Steve M. changed this in LFRic trunk a few weeks ago ([LFRic ticket 2397](https://code.metoffice.gov.uk/trac/lfric/ticket/2397)) so this PR brings PSyclone rules (mainly kernel stub generation) up to date.

**Comment by rupertford** [comment]
> > Hi @TeranIvy, I thought that if basic types were passed in to a subroutine and were only written then they should be intent(out). What am I missing?
> 
> @rupertford, the whole of a field and operator arrays are passed into a kernel, but the kernel only modifies one of the columns. Some compilers with some compile options will overwrite the full array of `intent(out)`, so the intent needs to be `inout` even for `GH_WRITE`. Steve M. changed this in LFRic trunk a few weeks ago ([LFRic ticket 2397](https://code.metoffice.gov.uk/trac/lfric/ticket/2397)) so this PR brings PSyclone rules (mainly kernel stub generation) up to date.

Ah, yes that makes sense. We're only writing to some of the data. Thanks.

**Review by arporter — CHANGES_REQUESTED** [review]
Good work Iva. There are just a couple of small changes requested in line.
Updated documentation builds without problems.
All tests are OK, incl. compilation. Ditto examples.
Changes are covered.
Please could you bring the branch up-to-date with master.
I've checked for pycodestyle/pylint - see inline.

File: src/psyclone/dynamo0p3.py [hunk 1]
```diff
@@ -8640,21 +8640,17 @@ def intent(self):
...
             return "inout"
         else:
-            valid_reductions = AccessType.get_valid_reduction_names()
+            valid_accesses = [AccessType.READ.api_specific_name()] + \
+                [access.api_specific_name() for access in write_accesses]
             raise GenerationError(
-                "Expecting argument access to be one of 'gh_read, gh_write, "
-                "gh_inc', 'gh_readwrite' or one of {0}, but found '{1}'".
-                format(valid_reductions, self.access))
+                "In the LFRic API argument access must be one of {0}, but "
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:8652]: Done and updated the relevant tests.

File: src/psyclone/dynamo0p3.py [hunk 2]
```diff
@@ -8640,21 +8640,17 @@ def intent(self):
...
+        write_accesses = AccessType.all_write_accesses()
         if self.access == AccessType.READ:
             return "in"
-        elif self.access == AccessType.WRITE:
-            return "out"
-        elif self.access == AccessType.READWRITE:
-            return "inout"
-        elif self.access in [AccessType.INC] + \
-                AccessType.get_valid_reduction_modes():
+        elif self.access in write_accesses:
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:8646]: Changed to `if`.

File: src/psyclone/dynamo0p3.py [hunk 3]
```diff
@@ -8640,21 +8640,17 @@ def intent(self):
...
             return "inout"
         else:
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:8648]: Removed the `else` and left the `GenerationError` for an invalid access.

**Comment by TeranIvy** [comment]
Addressed the review comments and brought the branch up-to-date with master.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Examples and tests still OK.
Docs still OK.

**Pull request merged**